### ISSUE TITLE
Segfault in ik command

### ISSUE THREAD

**Issue by adithyamurali** [mjd3/tracikpy#2]
Hi @mjd3 

Thanks for this amazing repo. I get an error when running the following test script in my Ubuntu 18.04 machine:

```
import numpy as np
from tracikpy import TracIKSolver

ee_pose = np.array([[ 0.0525767 , -0.64690764, -0.7607537 , 0.        ],
                    [-0.90099786, -0.35923817,  0.24320937, 0.2       ],
                    [-0.43062577,  0.67265031, -0.60174996, 0.4       ],
                    [ 0.        ,  0.        ,  0.        , 1.        ]])

ik_solver = TracIKSolver(
    "data/franka_panda.urdf",
    "panda_link0",
    "panda_hand",
)
qout = ik_solver.ik(ee_pose, qinit=np.zeros(ik_solver.number_of_joints))
print(qout)
```

There is a segfault at line `ik_solver.ik`:
`KDL FKSolver is failing: Segmentation fault (core dumped)`

I was wondering if you have seen this error before?

**Comment by adithyamurali** [comment]
I installed the package with the steps in the readme:

```
sudo apt-get install libeigen3-dev liborocos-kdl-dev libkdl-parser-dev liburdf-dev libnlopt-dev
git clone https://github.com/mjd3/tracikpy.git
pip install tracikpy/
```

**Comment by adithyamurali** [comment]
I believe the error is happening in these lines:
https://github.com/mjd3/tracikpy/blob/main/tracikpy/src/nlopt_ik.cpp#L354
https://github.com/mjd3/tracikpy/blob/main/tracikpy/src/nlopt_ik.cpp#L407

**Comment by mjd3** [comment]
Hmmm, I am not able to reproduce this error. I have added an Ubuntu 18.04 workflow to testing as well as Dockerfiles for Ubuntu 18.04 and Ubuntu 20.04 and all of these pass the unit tests without issue (see #3). Can you install with the `test` option (using `pip install tracikpy[test]`) and then run `pytest` from the root directory of the project to see if that passes?

**Comment by adithyamurali** [comment]
Thanks for the debuging steps. I have a small correction: actually in the first post above, I installed as `pip install trackikpy[test]` and that's how I ended up with the segfault error.

If I just install with `pip install tracikpy/` without the `test` option, I get a separate error: `ModuleNotFoundError: No module named tracikpy.swig.trac_ik_wrap` when trying to import the library: `from tracikpy import TracIKSolver`

**Comment by mjd3** [comment]
Ah, this suggests that the install is not working. Can you try with `pip install tracikpy --verbose` and see if the SWIG binding fails to build? Also, make sure you have SWIG installed (on Ubuntu, you can check with `swig -version` from the command line).

**Comment by adithyamurali** [comment]
Yes, it seems I do have swig installed, here's the output of `swig -version`:

```
SWIG Version 3.0.12
Compiled with g++ [x86_64-pc-linux-gnu]
Configured options: +pcre
Please see http://www.swig.org for reporting bugs and further information
```

When I installed with the `--verbose` option, there were no build errors. Here's the full output:

`Using pip 21.1.3 from /home/admurali/anaconda3/envs/rlgpu/lib/python3.7/site-packages/pip (python 3.7)
Non-user install because site-packages writeable
Created temporary directory: /tmp/pip-ephem-wheel-cache-wdc8wzmh
Created temporary directory: /tmp/pip-req-tracker-ulmy33cn
Initialized build tracking at /tmp/pip-req-tracker-ulmy33cn
Created build tracker: /tmp/pip-req-tracker-ulmy33cn
Entered build tracker: /tmp/pip-req-tracker-ulmy33cn
Created temporary directory: /tmp/pip-install-guvp44w5
Processing ./tracikpy
  Created temporary directory: /tmp/pip-req-build-168qvv_i
  DEPRECATION: A future pip version will change local packages to be built in-place without first copying to a temporary directory. We recommend you use --use-feature=in-tree-build to test your packages with this new behavior before it becomes the default.
   pip 21.3 will remove support for this functionality. You can find discussion regarding this at https://github.com/pypa/pip/issues/7555.
  Added file:///home/admurali/ws/tracikpy to build tracker '/tmp/pip-req-tracker-ulmy33cn'
    Running setup.py (path:/tmp/pip-req-build-168qvv_i/setup.py) egg_info for package from file:///home/admurali/ws/tracikpy
    Created temporary directory: /tmp/pip-pip-egg-info-eooyqyjb
    Running command python setup.py egg_info
    running egg_info
    creating /tmp/pip-pip-egg-info-eooyqyjb/tracikpy.egg-info
    writing /tmp/pip-pip-egg-info-eooyqyjb/tracikpy.egg-info/PKG-INFO
    writing dependency_links to /tmp/pip-pip-egg-info-eooyqyjb/tracikpy.egg-info/dependency_links.txt
    writing requirements to /tmp/pip-pip-egg-info-eooyqyjb/tracikpy.egg-info/requires.txt
    writing top-level names to /tmp/pip-pip-egg-info-eooyqyjb/tracikpy.egg-info/top_level.txt
    writing manifest file '/tmp/pip-pip-egg-info-eooyqyjb/tracikpy.egg-info/SOURCES.txt'
    reading manifest file '/tmp/pip-pip-egg-info-eooyqyjb/tracikpy.egg-info/SOURCES.txt'
    writing manifest file '/tmp/pip-pip-egg-info-eooyqyjb/tracikpy.egg-info/SOURCES.txt'
  Source in /tmp/pip-req-build-168qvv_i has version 0.1.0, which satisfies requirement tracikpy==0.1.0 from file:///home/admurali/ws/tracikpy
  Removed tracikpy==0.1.0 from file:///home/admurali/ws/tracikpy from build tracker '/tmp/pip-req-tracker-ulmy33cn'
Requirement already satisfied: numpy in /home/admurali/anaconda3/envs/rlgpu/lib/python3.7/site-packages (from tracikpy==0.1.0) (1.21.1)
Created temporary directory: /tmp/pip-unpack-_ui5goh_
Building wheels for collected packages: tracikpy
  Created temporary directory: /tmp/pip-wheel-2z11bl_7
  Building wheel for tracikpy (setup.py) ...   Destination directory: /tmp/pip-wheel-2z11bl_7
  Running command /home/admurali/anaconda3/envs/rlgpu/bin/python3.7 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-168qvv_i/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-168qvv_i/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-2z11bl_7
  running bdist_wheel
  running build
  running build_py
  running build_ext
  building 'tracikpy.swig._trac_ik_wrap' extension
  swigging tracikpy/swig/trac_ik.i to tracikpy/swig/trac_ik_wrap.cpp
  swig -python -c++ -Itracikpy/include -o tracikpy/swig/trac_ik_wrap.cpp tracikpy/swig/trac_ik.i
  creating build
  creating build/temp.linux-x86_64-3.7
  creating build/temp.linux-x86_64-3.7/tracikpy
  creating build/temp.linux-x86_64-3.7/tracikpy/swig
  creating build/temp.linux-x86_64-3.7/tracikpy/src
  gcc -pthread -B /home/admurali/anaconda3/envs/rlgpu/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -Itracikpy/include -I/usr/include/eigen3 -I/home/admurali/anaconda3/envs/rlgpu/include/python3.7m -c tracikpy/swig/trac_ik_wrap.cpp -o build/temp.linux-x86_64-3.7/tracikpy/swig/trac_ik_wrap.o
  cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
  gcc -pthread -B /home/admurali/anaconda3/envs/rlgpu/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -Itracikpy/include -I/usr/include/eigen3 -I/home/admurali/anaconda3/envs/rlgpu/include/python3.7m -c tracikpy/src/trac_ik.cpp -o build/temp.linux-x86_64-3.7/tracikpy/src/trac_ik.o
  cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
  gcc -pthread -B /home/admurali/anaconda3/envs/rlgpu/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -Itracikpy/include -I/usr/include/eigen3 -I/home/admurali/anaconda3/envs/rlgpu/include/python3.7m -c tracikpy/src/nlopt_ik.cpp -o build/temp.linux-x86_64-3.7/tracikpy/src/nlopt_ik.o
  cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
  tracikpy/src/nlopt_ik.cpp: In member function ‘int NLOPT_IK::NLOPT_IK::CartToJnt(const KDL::JntArray&, const KDL::Frame&, KDL::JntArray&, KDL::Twist, const KDL::JntArray&)’:
  tracikpy/src/nlopt_ik.cpp:467:26: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     if (q_init.data.size() != types.size())
         ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
  gcc -pthread -B /home/admurali/anaconda3/envs/rlgpu/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -Itracikpy/include -I/usr/include/eigen3 -I/home/admurali/anaconda3/envs/rlgpu/include/python3.7m -c tracikpy/src/kdl_tl.cpp -o build/temp.linux-x86_64-3.7/tracikpy/src/kdl_tl.o
  cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
  creating build/lib.linux-x86_64-3.7
  creating build/lib.linux-x86_64-3.7/tracikpy
  creating build/lib.linux-x86_64-3.7/tracikpy/swig
  g++ -pthread -shared -B /home/admurali/anaconda3/envs/rlgpu/compiler_compat -L/home/admurali/anaconda3/envs/rlgpu/lib -Wl,-rpath=/home/admurali/anaconda3/envs/rlgpu/lib -Wl,--no-as-needed -Wl,--sysroot=/ build/temp.linux-x86_64-3.7/tracikpy/swig/trac_ik_wrap.o build/temp.linux-x86_64-3.7/tracikpy/src/trac_ik.o build/temp.linux-x86_64-3.7/tracikpy/src/nlopt_ik.o build/temp.linux-x86_64-3.7/tracikpy/src/kdl_tl.o -lorocos-kdl -lnlopt -lurdf -lkdl_parser -o build/lib.linux-x86_64-3.7/tracikpy/swig/_trac_ik_wrap.cpython-37m-x86_64-linux-gnu.so
  copying tracikpy/trac_ik_solver.py -> build/lib.linux-x86_64-3.7/tracikpy
  copying tracikpy/version.py -> build/lib.linux-x86_64-3.7/tracikpy
  copying tracikpy/__init__.py -> build/lib.linux-x86_64-3.7/tracikpy
  copying tracikpy/swig/trac_ik_wrap.py -> build/lib.linux-x86_64-3.7/tracikpy/swig
  copying tracikpy/swig/__init__.py -> build/lib.linux-x86_64-3.7/tracikpy/swig
  installing to build/bdist.linux-x86_64/wheel
  running install
  running install_lib
  creating build/bdist.linux-x86_64
  creating build/bdist.linux-x86_64/wheel
  creating build/bdist.linux-x86_64/wheel/tracikpy
  creating build/bdist.linux-x86_64/wheel/tracikpy/swig
  copying build/lib.linux-x86_64-3.7/tracikpy/swig/_trac_ik_wrap.cpython-37m-x86_64-linux-gnu.so -> build/bdist.linux-x86_64/wheel/tracikpy/swig
  copying build/lib.linux-x86_64-3.7/tracikpy/swig/trac_ik_wrap.py -> build/bdist.linux-x86_64/wheel/tracikpy/swig
  copying build/lib.linux-x86_64-3.7/tracikpy/swig/__init__.py -> build/bdist.linux-x86_64/wheel/tracikpy/swig
  copying build/lib.linux-x86_64-3.7/tracikpy/trac_ik_solver.py -> build/bdist.linux-x86_64/wheel/tracikpy
  copying build/lib.linux-x86_64-3.7/tracikpy/version.py -> build/bdist.linux-x86_64/wheel/tracikpy
  copying build/lib.linux-x86_64-3.7/tracikpy/__init__.py -> build/bdist.linux-x86_64/wheel/tracikpy
  running install_egg_info
  running egg_info
  creating tracikpy.egg-info
  writing tracikpy.egg-info/PKG-INFO
  writing dependency_links to tracikpy.egg-info/dependency_links.txt
  writing requirements to tracikpy.egg-info/requires.txt
  writing top-level names to tracikpy.egg-info/top_level.txt
  writing manifest file 'tracikpy.egg-info/SOURCES.txt'
  reading manifest file 'tracikpy.egg-info/SOURCES.txt'
  writing manifest file 'tracikpy.egg-info/SOURCES.txt'
  Copying tracikpy.egg-info to build/bdist.linux-x86_64/wheel/tracikpy-0.1.0-py3.7.egg-info
  running install_scripts
  adding license file "LICENSE" (matched pattern "LICEN[CS]E*")
  creating build/bdist.linux-x86_64/wheel/tracikpy-0.1.0.dist-info/WHEEL
  creating '/tmp/pip-wheel-2z11bl_7/tracikpy-0.1.0-cp37-cp37m-linux_x86_64.whl' and adding 'build/bdist.linux-x86_64/wheel' to it
  adding 'tracikpy/__init__.py'
  adding 'tracikpy/trac_ik_solver.py'
  adding 'tracikpy/version.py'
  adding 'tracikpy/swig/__init__.py'
  adding 'tracikpy/swig/_trac_ik_wrap.cpython-37m-x86_64-linux-gnu.so'
  adding 'tracikpy/swig/trac_ik_wrap.py'
  adding 'tracikpy-0.1.0.dist-info/LICENSE'
  adding 'tracikpy-0.1.0.dist-info/METADATA'
  adding 'tracikpy-0.1.0.dist-info/WHEEL'
  adding 'tracikpy-0.1.0.dist-info/top_level.txt'
  adding 'tracikpy-0.1.0.dist-info/RECORD'
  removing build/bdist.linux-x86_64/wheel
done
  Created wheel for tracikpy: filename=tracikpy-0.1.0-cp37-cp37m-linux_x86_64.whl size=3115277 sha256=bca5f73733ee383684b581a020fdf9891480666db2910afabd018d98239412c2
  Stored in directory: /tmp/pip-ephem-wheel-cache-wdc8wzmh/wheels/5f/2d/2f/182aa95356f03e834b6b0100ad29fd64ca37fcb953f587b3b7
Successfully built tracikpy
Installing collected packages: tracikpy

Successfully installed tracikpy-0.1.0
Removed build tracker: '/tmp/pip-req-tracker-ulmy33cn'`

**Comment by mjd3** [comment]
Hmm yes it does seem like the install is working correctly. So after this successful install you are unable to import?

**Comment by adithyamurali** [comment]
Yeah after this installation, here's the error trace i get when running `from tracikpy import TracIKSolver`:

```
<ipython-input-1-4093a645420a> in <module>()
      1 import numpy as np
----> 2 from tracikpy import TracIKSolver
      3 
      4 ee_pose = np.array([[ 0.0525767 , -0.64690764, -0.7607537 , 0.        ],
      5                     [-0.90099786, -0.35923817,  0.24320937, 0.2       ],

/home/admurali/ws/tracikpy/tracikpy/__init__.py in <module>()
----> 1 from .trac_ik_solver import TracIKSolver

/home/admurali/ws/tracikpy/tracikpy/trac_ik_solver.py in <module>()
      7 import numpy as np
      8 
----> 9 from .swig.trac_ik_wrap import TRAC_IK
     10 
     11 

ModuleNotFoundError: No module named 'tracikpy.swig.trac_ik_wrap'

```

**Comment by mjd3** [comment]
Are you installing with `pip install -e`? Does that file (`tracikpy/swig/trac_ik_wrap.py`) exist after building the swig bindings? Sorry for the issues! Hoping we can get this figured out although it is hard to debug something of this nature if I am unable to reproduce myself (e.g., in docker).

**Comment by mjd3** [comment]
@adithyamurali were you able to resolve?
